File: 993/993.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/993.py

val PERIOD_START: Long = 514L
val PERIOD: Long = 71L
val DELTA_PATTERN: LongArray = longArrayOf(
    17, -2, -8, -2, -2, -14, -2, -2, -17, -8, -5, -8, -5, -2, -2, -5, -8,
    50, -8, 23, -13, -2, 67, -5, -2, -2, -5, -8, -5, 21, 29, -11, -2, -2,
    6, -11, 31, -2, -11, 17, -2, -8, -2, -2, -14, -2, -2, -17, -8, -5, -8,
    -8, 8, -13, -5, -2, -2, -5, -2, -11, -8, -8, -5, -2, -11, -8, -8, -5,
    -2, -11, 216,
)
val PATTERN_SUM: Long = DELTA_PATTERN.sum()

data class GameState(val pos: Long, val carry: Long, val bananas: Set<Long>)

fun stepState(pos: Long, carry: Long, bananas: Set<Long>): GameState? {
    val hasX = bananas.contains(pos)
    val hasX1 = bananas.contains(pos + 1L)
    if (hasX && hasX1) {
        val nb = bananas.toMutableSet()
        nb.remove(pos + 1L)
        return GameState(pos - 1L, carry + 1L, nb)
    }
    if (hasX && !hasX1) {
        val nb = bananas.toMutableSet()
        nb.remove(pos)
        return GameState(pos + 2L, carry + 1L, nb)
    }
    if (!hasX && hasX1) {
        val nb = bananas.toMutableSet()
        nb.remove(pos + 1L)
        nb.add(pos)
        return GameState(pos + 2L, carry, nb)
    }
    if (carry >= 3L) {
        val nb = bananas.toMutableSet()
        nb.add(pos - 1L)
        nb.add(pos)
        nb.add(pos + 1L)
        return GameState(pos - 2L, carry - 3L, nb)
    }
    return null
}

fun simulateSteps(initialBananas: Long, steps: Int): GameState {
    var pos = 0L
    var carry = initialBananas
    var bananas: Set<Long> = emptySet()
    for (i in 0 until steps) {
        val nxt = stepState(pos, carry, bananas) ?: break
        pos = nxt.pos; carry = nxt.carry; bananas = nxt.bananas
    }
    return GameState(pos, carry, bananas)
}

fun simulateBbValues(limit: Long): LongArray {
    val bb = mutableListOf(0L)
    var pos = 0L
    var carry = 0L
    var bananas: Set<Long> = emptySet()
    for (n in 1L..limit) {
        carry += 1L
        while (true) {
            val nxt = stepState(pos, carry, bananas)
            if (nxt == null) { bb.add(pos); break }
            pos = nxt.pos; carry = nxt.carry; bananas = nxt.bananas
        }
    }
    return bb.toLongArray()
}

fun buildPrefix(): LongArray = simulateBbValues(PERIOD_START + PERIOD)

fun bbFunc(n: Long, bbPrefix: LongArray): Long {
    if (n <= PERIOD_START) return bbPrefix[n.toInt()]
    val remaining = n - PERIOD_START
    val wholePeriods = remaining / PERIOD
    val tail = (remaining % PERIOD).toInt()
    return bbPrefix[PERIOD_START.toInt()] + wholePeriods * PATTERN_SUM + DELTA_PATTERN.take(tail).sum()
}

fun main() {
    val bbPrefix = buildPrefix()

    val s1 = simulateSteps(3L, 1)
    check(s1.pos == -2L)
    check(s1.carry == 0L)
    check(s1.bananas == setOf(-1L, 0L, 1L))

    val s2 = simulateSteps(5L, 5)
    check(s2.pos == -1L)
    check(s2.carry == 0L)
    check(s2.bananas == setOf(-2L, -1L, 0L, 1L, 2L))

    check(bbFunc(1000L, bbPrefix) == 1499L)

    val deltas = LongArray(bbPrefix.size - 1) { i -> bbPrefix[i + 1] - bbPrefix[i] }
    for (i in DELTA_PATTERN.indices) {
        check(deltas[PERIOD_START.toInt() + i] == DELTA_PATTERN[i])
    }

    println(bbFunc(1_000_000_000_000_000_000L, bbPrefix))
}

File: 993/993.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/993.py

use std::collections::HashSet;

const PERIOD_START: i64 = 514;
const PERIOD: i64 = 71;
const DELTA_PATTERN: [i64; 71] = [
    17, -2, -8, -2, -2, -14, -2, -2, -17, -8, -5, -8, -5, -2, -2, -5, -8, 50, -8, 23, -13, -2,
    67, -5, -2, -2, -5, -8, -5, 21, 29, -11, -2, -2, 6, -11, 31, -2, -11, 17, -2, -8, -2, -2,
    -14, -2, -2, -17, -8, -5, -8, -8, 8, -13, -5, -2, -2, -5, -2, -11, -8, -8, -5, -2, -11, -8,
    -8, -5, -2, -11, 216,
];

fn pattern_sum() -> i64 {
    DELTA_PATTERN.iter().sum()
}

fn step_state(
    pos: i64,
    carry: i64,
    bananas: &HashSet<i64>,
) -> Option<(i64, i64, HashSet<i64>)> {
    let has_x = bananas.contains(&pos);
    let has_x1 = bananas.contains(&(pos + 1));
    if has_x && has_x1 {
        let mut new_bananas = bananas.clone();
        new_bananas.remove(&(pos + 1));
        return Some((pos - 1, carry + 1, new_bananas));
    }
    if has_x && !has_x1 {
        let mut new_bananas = bananas.clone();
        new_bananas.remove(&pos);
        return Some((pos + 2, carry + 1, new_bananas));
    }
    if !has_x && has_x1 {
        let mut new_bananas = bananas.clone();
        new_bananas.remove(&(pos + 1));
        new_bananas.insert(pos);
        return Some((pos + 2, carry, new_bananas));
    }
    if carry >= 3 {
        let mut new_bananas = bananas.clone();
        new_bananas.insert(pos - 1);
        new_bananas.insert(pos);
        new_bananas.insert(pos + 1);
        return Some((pos - 2, carry - 3, new_bananas));
    }
    None
}

fn simulate_steps(initial_bananas: i64, steps: usize) -> (i64, i64, HashSet<i64>) {
    let mut pos: i64 = 0;
    let mut carry = initial_bananas;
    let mut bananas: HashSet<i64> = HashSet::new();
    for _ in 0..steps {
        match step_state(pos, carry, &bananas) {
            None => break,
            Some((np, nc, nb)) => {
                pos = np;
                carry = nc;
                bananas = nb;
            }
        }
    }
    (pos, carry, bananas)
}

fn simulate_bb_values(limit: usize) -> Vec<i64> {
    let mut bb = vec![0i64];
    let mut pos: i64 = 0;
    let mut carry: i64 = 0;
    let mut bananas: HashSet<i64> = HashSet::new();
    for _ in 1..=(limit as i64) {
        carry += 1;
        loop {
            match step_state(pos, carry, &bananas) {
                None => {
                    bb.push(pos);
                    break;
                }
                Some((np, nc, nb)) => {
                    pos = np;
                    carry = nc;
                    bananas = nb;
                }
            }
        }
    }
    bb
}

fn build_prefix() -> Vec<i64> {
    simulate_bb_values((PERIOD_START + PERIOD) as usize)
}

fn bb_func(n: i64, bb_prefix: &[i64]) -> i64 {
    if n <= PERIOD_START {
        return bb_prefix[n as usize];
    }
    let remaining = n - PERIOD_START;
    let whole_periods = remaining / PERIOD;
    let tail = (remaining % PERIOD) as usize;
    bb_prefix[PERIOD_START as usize]
        + whole_periods * pattern_sum()
        + DELTA_PATTERN[..tail].iter().sum::<i64>()
}

fn main() {
    let bb_prefix = build_prefix();

    let (pos, carry, bananas) = simulate_steps(3, 1);
    assert_eq!(pos, -2);
    assert_eq!(carry, 0);
    let expected1: HashSet<i64> = [-1, 0, 1].iter().cloned().collect();
    assert_eq!(bananas, expected1);

    let (pos2, carry2, bananas2) = simulate_steps(5, 5);
    assert_eq!(pos2, -1);
    assert_eq!(carry2, 0);
    let expected2: HashSet<i64> = [-2, -1, 0, 1, 2].iter().cloned().collect();
    assert_eq!(bananas2, expected2);

    assert_eq!(bb_func(1000, &bb_prefix), 1499);

    let deltas: Vec<i64> = (0..bb_prefix.len() - 1)
        .map(|i| bb_prefix[i + 1] - bb_prefix[i])
        .collect();
    for (i, &delta) in DELTA_PATTERN.iter().enumerate() {
        assert_eq!(deltas[PERIOD_START as usize + i], delta);
    }

    println!("{}", bb_func(1_000_000_000_000_000_000i64, &bb_prefix));
}
